### DISCUSSION TITLE
[Feature Request] Add a Read-only feature to fields

### DISCUSSION THREAD

**Discussion by ghost** [craftcms/cms#8636] [Ideas]
### Description
I think Craft would be more awesome and more handsome if it allowed me to mark some fields as read-only. just a small checkbox in the field's settings 😽

**Comment by carlcs** [comment]
There’s a plugin for it https://plugins.craftcms.com/incognito-field

**Comment by brandonkelly** [comment]
@Sam-Black-0013 Is that plugin relatively similar to what you’re looking for?

**Comment by ghost** [comment]
Yub it is, thanks. But can't it be added as a core feature in fields ?

**Comment by brandonkelly** [comment]
@Sam-Black-0013 Just making sure we’re on the same page. There are lots of similar enhancements planned for Craft 4.0 so I will make a note of this for that.

**Comment by ghost** [comment]
AWW @brandonkelly u guys are the best >.<

**Comment by missmatsuko** [comment]
I would find it handy if there could be readonly setting on fields by user roles. Like if admins/devs could edit the field from the CMS without changing the field type from readonly to editable, but it would always be readonly for everyone else.

**Comment by brandonkelly** [comment]
Craft 3.5 has a new field layout designer, which includes the ability to add additional UI elements, including one that pulls in a site template of your choosing. So you can use that to put whatever arbitrary HTML you want in your field layout, which I’d say resolves this.

**Comment by carlcs** [comment]
Brandon, I think you misunderstood the feature request. The idea is to be able to configure regular fields’ visibility on edit pages. Ideally this would work with every field type, and configuration would be per field layout.

The Incognito field plugin solves this for plain text fields only. Visibility can be configured dynamically using Twig.

```twig
{{ currentUser.isInGroup('editors') ? 'readonly' : 'hidden' }}
```

https://github.com/mmikkel/IncognitoField-Craft3#overriding-the-rendering-mode-with-twig-logic

**Comment by brandonkelly** [comment]
There hasn’t been any mention of toggling field _visibility_ here, just _writability_. Toggling field visibility based on user groups is covered by #947.

I interpreted the issue as requesting a _field-level_ “Read-only” checkbox – not something that could sometimes be writable, sometimes not, depending on a condition. At which point, it would be better to just not have a field in the first place, and use a Template UI element in the field layout. But can reopen if I was mistaken.

**Comment by carlcs** [comment]
Ah ok, that’s how you read it! I don’t see the purpose of a field that’s not writeable at all, so I assumed the issue is about its rendering in the control panel. Content would be written to it in other ways, from front end forms, or programatically.

#947 is pretty much covering my intepretetion of this request.

**Comment by brandonkelly** [comment]
> I don’t see the purpose of a field that’s not writeable at all

Yeah good point. I’ll go ahead and reopen on the assumption that the request is for _sometimes_-read-only fields.

**Comment by sander-lameco** [comment]
We are also looking for this feature.

In some projects we have data that is coming from an external source into Craft using Feed Me. Now we hide those fields for certain User Groups so they would not accidentally edit the data in Craft and after a new feed run gets overwritten.

Now we would like the ability to add a Readonly condition so the fields are still visible but not editable for certain User Groups. That way the data is still viewable but they won't be able to manipulate it in Craft.

**Reply by thupsi** [reply]
@sander-lameco A way to achieve this with current native functionality would be to add to the field layout a custom UI element that simply outputs the field's value. You could then display this UI element to any user groups you like.

**Reply by sander-lameco** [reply]
Would we have to do this for every field separately (or a combination of all fields).
Of course it would be nicer if it is a build in feature because else if a field(Handle) would change we would have to adjust the Custom UI element template as wel.

**Comment by khalwat** [comment]
This came up with regards to my Code Field plugin, FWIW:

https://github.com/nystudio107/craft-code-field/issues/15

Rather than implement it just for this field, I do think it would be useful to be baked into Craft at some point, with user-level permissions for the field.

Craft could use the `Field::getStaticHtml()` method that already exists to render fields that are read-only.

**Comment by khalwat** [comment]
Related: https://github.com/craftcms/cms/discussions/14015

**Comment by regularlabs** [comment]
In the Field Layout (of an Entry Type) you can set a field to be required.
Why not simply add an option there to make it read-only?

![image](https://github.com/craftcms/cms/assets/13553242/9b667ea6-8181-4d10-8074-fefe67c85c28)

**Comment by JayBox325** [comment]
I'm interested in this being added to core Craft, is there any plans for this to be added?
